pandas/core/tools/datetimes.py

```diff
@@ -557,7 +559,48 @@ def _to_datetime_with_unit(arg, unit, name, utc: bool, errors: str) -> Index:
         tz_parsed = None
     else:
         arg = np.asarray(arg)
-        arr, tz_parsed = tslib.array_with_unit_to_datetime(arg, unit, errors=errors)
+
+        if arg.dtype.kind in ["i", "u"]:
+            # Note we can't do "f" here because that could induce unwanted
+            #  rounding GH#14156, GH#20445
+            arr = arg.astype(f"datetime64[{unit}]", copy=False)
+            try:
+                arr = astype_overflowsafe(arr, np.dtype("M8[ns]"), copy=False)
+            except OutOfBoundsDatetime:
+                if errors == "raise":
+                    raise
+                arg = arg.astype(object)
+                return _to_datetime_with_unit(arg, unit, name, utc, errors)
```

Comment:
we don't need to define mask explicitly here bc it is understood by astype_overflowsafe

Reply:
ah right, this would hit it

https://github.com/pandas-dev/pandas/blob/1d5f05c33c613508727ee7b971ad56723d474446/pandas/tests/tools/test_to_datetime.py#L1570-L1574

all good then

```python
(Pdb) arg
array([          1370745748,           1370745749,           1370745750,
                 1370745751,           1370745752,           1370745753,
                 1370745754,           1370745755,           1370745756,
                 1370745757,           1370745758,           1370745759,
                 1370745760,           1370745761,           1370745762,
                 1370745763,           1370745764,           1370745765,
                 1370745766,           1370745767, -9223372036854775808])
(Pdb) astype_overflowsafe(arr, np.dtype("M8[ns]"), copy=False)
array(['2013-06-09T02:42:28.000000000', '2013-06-09T02:42:29.000000000',
       '2013-06-09T02:42:30.000000000', '2013-06-09T02:42:31.000000000',
       '2013-06-09T02:42:32.000000000', '2013-06-09T02:42:33.000000000',
       '2013-06-09T02:42:34.000000000', '2013-06-09T02:42:35.000000000',
       '2013-06-09T02:42:36.000000000', '2013-06-09T02:42:37.000000000',
       '2013-06-09T02:42:38.000000000', '2013-06-09T02:42:39.000000000',
       '2013-06-09T02:42:40.000000000', '2013-06-09T02:42:41.000000000',
       '2013-06-09T02:42:42.000000000', '2013-06-09T02:42:43.000000000',
       '2013-06-09T02:42:44.000000000', '2013-06-09T02:42:45.000000000',
       '2013-06-09T02:42:46.000000000', '2013-06-09T02:42:47.000000000',
                                 'NaT'], dtype='datetime64[ns]')
```